{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "from sentence_transformers import SentenceTransformer\n",
    "\n",
    "model = SentenceTransformer(\"sentence-transformers/gtr-t5-large\")"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 1. Prepare data\n",
    "\n",
    "I will prototype this approach by extracting the relations between a subset of the Open Targets data. The subset will be the following:\n",
    "- Molecules. \n",
    "- Drug indications.\n",
    "- Drug mechanism of actions.\n",
    "- Diseases.\n",
    "\n",
    "I am going to evaluate the model by the ability to find these relationships:\n",
    "- `drug_df.linkedDiseases` is a list of `disease_df.id` values.\n",
    "- `drug_df.linkedTargets` is a list of `target_df.id` values.\n",
    "- `indication_df.approvedIndications` is a list of `disease_df.id` values.\n",
    "- `indication_df.id` is a list of `drug_df.id` values.\n",
    "- `moa_df.targets` is a list of `target_df.id` values.\n",
    "- `moa_df.chemblIds` is a list of `drug_df.id` values."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Download data from 22.11 release\n",
    "# !wget --recursive --no-parent --no-host-directories --cut-dirs 8 ftp://ftp.ebi.ac.uk/pub/databases/opentargets/platform/22.11/output/etl/parquet/molecule --accept \"*.parquet\" --directory-prefix data\n",
    "# !wget --recursive --no-parent --no-host-directories --cut-dirs 8 ftp://ftp.ebi.ac.uk/pub/databases/opentargets/platform/22.11/output/etl/parquet/indication --accept \"*.parquet\" --directory-prefix data\n",
    "# !wget --recursive --no-parent --no-host-directories --cut-dirs 8 ftp://ftp.ebi.ac.uk/pub/databases/opentargets/platform/22.11/output/etl/parquet/mechanismOfAction --accept \"*.parquet\" --directory-prefix data\n",
    "# !wget --recursive --no-parent --no-host-directories --cut-dirs 8 ftp://ftp.ebi.ac.uk/pub/databases/opentargets/platform/22.11/output/etl/parquet/diseases --accept \"*.parquet\" --directory-prefix data\n",
    "# !wget --recursive --no-parent --no-host-directories --cut-dirs 8 ftp://ftp.ebi.ac.uk/pub/databases/opentargets/platform/22.11/output/etl/parquet/targets --accept \"*.parquet\" --directory-prefix data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "\n",
    "drug_df = pd.concat([pd.read_parquet(f\"data/molecule/{f}\") for f in os.listdir(\"data/molecule\")])\n",
    "indication_df = pd.concat([pd.read_parquet(f\"data/indication/{f}\") for f in os.listdir(\"data/indication\")])\n",
    "moa_df = pd.concat([pd.read_parquet(f\"data/mechanismOfAction/{f}\") for f in os.listdir(\"data/mechanismOfAction\")])\n",
    "disease_df = pd.concat([pd.read_parquet(f\"data/diseases/{f}\") for f in os.listdir(\"data/diseases\")])\n",
    "target_df = pd.concat([pd.read_parquet(f\"data/targets/{f}\") for f in os.listdir(\"data/targets\")])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "Int64Index: 12854 entries, 0 to 59\n",
      "Data columns (total 19 columns):\n",
      " #   Column                     Non-Null Count  Dtype  \n",
      "---  ------                     --------------  -----  \n",
      " 0   id                         12854 non-null  object \n",
      " 1   canonicalSmiles            10260 non-null  object \n",
      " 2   inchiKey                   10260 non-null  object \n",
      " 3   drugType                   12845 non-null  object \n",
      " 4   blackBoxWarning            12845 non-null  object \n",
      " 5   name                       12854 non-null  object \n",
      " 6   yearOfFirstApproval        3103 non-null   float64\n",
      " 7   maximumClinicalTrialPhase  12845 non-null  float64\n",
      " 8   parentId                   1451 non-null   object \n",
      " 9   hasBeenWithdrawn           12845 non-null  object \n",
      " 10  isApproved                 12845 non-null  object \n",
      " 11  withdrawnNotice            232 non-null    object \n",
      " 12  tradeNames                 12854 non-null  object \n",
      " 13  synonyms                   12854 non-null  object \n",
      " 14  crossReferences            9338 non-null   object \n",
      " 15  childChemblIds             2705 non-null   object \n",
      " 16  linkedDiseases             8334 non-null   object \n",
      " 17  linkedTargets              6120 non-null   object \n",
      " 18  description                12845 non-null  object \n",
      "dtypes: float64(2), object(17)\n",
      "memory usage: 2.5+ MB\n",
      "None\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>canonicalSmiles</th>\n",
       "      <th>inchiKey</th>\n",
       "      <th>drugType</th>\n",
       "      <th>blackBoxWarning</th>\n",
       "      <th>name</th>\n",
       "      <th>yearOfFirstApproval</th>\n",
       "      <th>maximumClinicalTrialPhase</th>\n",
       "      <th>parentId</th>\n",
       "      <th>hasBeenWithdrawn</th>\n",
       "      <th>isApproved</th>\n",
       "      <th>withdrawnNotice</th>\n",
       "      <th>tradeNames</th>\n",
       "      <th>synonyms</th>\n",
       "      <th>crossReferences</th>\n",
       "      <th>childChemblIds</th>\n",
       "      <th>linkedDiseases</th>\n",
       "      <th>linkedTargets</th>\n",
       "      <th>description</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CHEMBL1096896</td>\n",
       "      <td>CCNC(=O)[C@H]1O[C@@H](n2cnc3c(NCC(c4ccccc4)c4c...</td>\n",
       "      <td>ZOTHAEBAWXWVID-HXEFRTELSA-N</td>\n",
       "      <td>Small molecule</td>\n",
       "      <td>False</td>\n",
       "      <td>UK432097</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "      <td>None</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>None</td>\n",
       "      <td>[]</td>\n",
       "      <td>[Uk432097]</td>\n",
       "      <td>[(drugbank, [DB12691])]</td>\n",
       "      <td>None</td>\n",
       "      <td>{'rows': ['EFO_0000341'], 'count': 1}</td>\n",
       "      <td>{'rows': ['ENSG00000128271'], 'count': 1}</td>\n",
       "      <td>Small molecule drug with a maximum clinical tr...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CHEMBL1200698</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Small molecule</td>\n",
       "      <td>False</td>\n",
       "      <td>PENTETATE CALCIUM TRISODIUM YB 169</td>\n",
       "      <td>1976.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>CHEMBL2110560</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>None</td>\n",
       "      <td>[Ytterbium yb 169 dtpa]</td>\n",
       "      <td>[169yb, COMPOUND 24266, MATERIAL A, MATERIAL-A...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>{'rows': [], 'count': 0}</td>\n",
       "      <td>Small molecule drug with a maximum clinical tr...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CHEMBL1200747</td>\n",
       "      <td>CC(O)C(=O)O.N</td>\n",
       "      <td>RZOBLYBZQXQGFY-UHFFFAOYSA-N</td>\n",
       "      <td>Small molecule</td>\n",
       "      <td>False</td>\n",
       "      <td>AMMONIUM LACTATE</td>\n",
       "      <td>1985.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>None</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>None</td>\n",
       "      <td>[Ammonium lactate, Lac-hydrin]</td>\n",
       "      <td>[Ammonium lactate, BMS-186091, E328, Lac hydri...</td>\n",
       "      <td>[(DailyMed, [ammonium%20lactate]), (DrugCentra...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>{'rows': [], 'count': 0}</td>\n",
       "      <td>Small molecule drug with a maximum clinical tr...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CHEMBL1200869</td>\n",
       "      <td>Cc1ncc([N+](=O)[O-])n1CCO.Cl</td>\n",
       "      <td>FPTPAIQTXYFGJC-UHFFFAOYSA-N</td>\n",
       "      <td>Small molecule</td>\n",
       "      <td>False</td>\n",
       "      <td>METRONIDAZOLE HYDROCHLORIDE</td>\n",
       "      <td>1980.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>CHEMBL137</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>None</td>\n",
       "      <td>[Flagyl, Flagyl i.v., Metronidazole hydrochlor...</td>\n",
       "      <td>[Metronidazole hcl, Metronidazole hydrochlorid...</td>\n",
       "      <td>[(chEBI, [50687])]</td>\n",
       "      <td>None</td>\n",
       "      <td>{'rows': ['EFO_0000574', 'EFO_0003102', 'EFO_0...</td>\n",
       "      <td>{'rows': [], 'count': 0}</td>\n",
       "      <td>Small molecule drug with a maximum clinical tr...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CHEMBL1201190</td>\n",
       "      <td>Cl.Cl.O=C(O)COCCN1CCN([C@H](c2ccccc2)c2ccc(Cl)...</td>\n",
       "      <td>PGLIUCLTXOYQMV-GHVWMZMZSA-N</td>\n",
       "      <td>Small molecule</td>\n",
       "      <td>False</td>\n",
       "      <td>LEVOCETIRIZINE DIHYDROCHLORIDE</td>\n",
       "      <td>2007.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>CHEMBL1201191</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>None</td>\n",
       "      <td>[Levocetirizine dihydrochloride, Xusal, Xyzal,...</td>\n",
       "      <td>[Cetirizine (r)-form dihydrochloride, Levoceti...</td>\n",
       "      <td>[(DailyMed, [levocetirizine%20dihydrochloride])]</td>\n",
       "      <td>None</td>\n",
       "      <td>{'rows': ['EFO_0007141', 'HP_0012735', 'HP_000...</td>\n",
       "      <td>{'rows': ['ENSG00000196639'], 'count': 1}</td>\n",
       "      <td>Small molecule drug with a maximum clinical tr...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              id                                    canonicalSmiles  \\\n",
       "0  CHEMBL1096896  CCNC(=O)[C@H]1O[C@@H](n2cnc3c(NCC(c4ccccc4)c4c...   \n",
       "1  CHEMBL1200698                                               None   \n",
       "2  CHEMBL1200747                                      CC(O)C(=O)O.N   \n",
       "3  CHEMBL1200869                       Cc1ncc([N+](=O)[O-])n1CCO.Cl   \n",
       "4  CHEMBL1201190  Cl.Cl.O=C(O)COCCN1CCN([C@H](c2ccccc2)c2ccc(Cl)...   \n",
       "\n",
       "                      inchiKey        drugType blackBoxWarning  \\\n",
       "0  ZOTHAEBAWXWVID-HXEFRTELSA-N  Small molecule           False   \n",
       "1                         None  Small molecule           False   \n",
       "2  RZOBLYBZQXQGFY-UHFFFAOYSA-N  Small molecule           False   \n",
       "3  FPTPAIQTXYFGJC-UHFFFAOYSA-N  Small molecule           False   \n",
       "4  PGLIUCLTXOYQMV-GHVWMZMZSA-N  Small molecule           False   \n",
       "\n",
       "                                 name  yearOfFirstApproval  \\\n",
       "0                            UK432097                  NaN   \n",
       "1  PENTETATE CALCIUM TRISODIUM YB 169               1976.0   \n",
       "2                    AMMONIUM LACTATE               1985.0   \n",
       "3         METRONIDAZOLE HYDROCHLORIDE               1980.0   \n",
       "4      LEVOCETIRIZINE DIHYDROCHLORIDE               2007.0   \n",
       "\n",
       "   maximumClinicalTrialPhase       parentId hasBeenWithdrawn isApproved  \\\n",
       "0                        2.0           None            False      False   \n",
       "1                        4.0  CHEMBL2110560            False       True   \n",
       "2                        4.0           None            False       True   \n",
       "3                        4.0      CHEMBL137            False       True   \n",
       "4                        4.0  CHEMBL1201191            False       True   \n",
       "\n",
       "  withdrawnNotice                                         tradeNames  \\\n",
       "0            None                                                 []   \n",
       "1            None                            [Ytterbium yb 169 dtpa]   \n",
       "2            None                     [Ammonium lactate, Lac-hydrin]   \n",
       "3            None  [Flagyl, Flagyl i.v., Metronidazole hydrochlor...   \n",
       "4            None  [Levocetirizine dihydrochloride, Xusal, Xyzal,...   \n",
       "\n",
       "                                            synonyms  \\\n",
       "0                                         [Uk432097]   \n",
       "1  [169yb, COMPOUND 24266, MATERIAL A, MATERIAL-A...   \n",
       "2  [Ammonium lactate, BMS-186091, E328, Lac hydri...   \n",
       "3  [Metronidazole hcl, Metronidazole hydrochlorid...   \n",
       "4  [Cetirizine (r)-form dihydrochloride, Levoceti...   \n",
       "\n",
       "                                     crossReferences childChemblIds  \\\n",
       "0                            [(drugbank, [DB12691])]           None   \n",
       "1                                               None           None   \n",
       "2  [(DailyMed, [ammonium%20lactate]), (DrugCentra...           None   \n",
       "3                                 [(chEBI, [50687])]           None   \n",
       "4   [(DailyMed, [levocetirizine%20dihydrochloride])]           None   \n",
       "\n",
       "                                      linkedDiseases  \\\n",
       "0              {'rows': ['EFO_0000341'], 'count': 1}   \n",
       "1                                               None   \n",
       "2                                               None   \n",
       "3  {'rows': ['EFO_0000574', 'EFO_0003102', 'EFO_0...   \n",
       "4  {'rows': ['EFO_0007141', 'HP_0012735', 'HP_000...   \n",
       "\n",
       "                               linkedTargets  \\\n",
       "0  {'rows': ['ENSG00000128271'], 'count': 1}   \n",
       "1                   {'rows': [], 'count': 0}   \n",
       "2                   {'rows': [], 'count': 0}   \n",
       "3                   {'rows': [], 'count': 0}   \n",
       "4  {'rows': ['ENSG00000196639'], 'count': 1}   \n",
       "\n",
       "                                         description  \n",
       "0  Small molecule drug with a maximum clinical tr...  \n",
       "1  Small molecule drug with a maximum clinical tr...  \n",
       "2  Small molecule drug with a maximum clinical tr...  \n",
       "3  Small molecule drug with a maximum clinical tr...  \n",
       "4  Small molecule drug with a maximum clinical tr...  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "Int64Index: 8334 entries, 0 to 27\n",
      "Data columns (total 4 columns):\n",
      " #   Column               Non-Null Count  Dtype \n",
      "---  ------               --------------  ----- \n",
      " 0   id                   8334 non-null   object\n",
      " 1   indications          8334 non-null   object\n",
      " 2   approvedIndications  8334 non-null   object\n",
      " 3   indicationCount      8334 non-null   int32 \n",
      "dtypes: int32(1), object(3)\n",
      "memory usage: 293.0+ KB\n",
      "None\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>indications</th>\n",
       "      <th>approvedIndications</th>\n",
       "      <th>indicationCount</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CHEMBL2146121</td>\n",
       "      <td>[{'disease': 'EFO_0000565', 'efoName': 'leukem...</td>\n",
       "      <td>[]</td>\n",
       "      <td>42</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CHEMBL2108738</td>\n",
       "      <td>[{'disease': 'EFO_0005922', 'efoName': 'esopha...</td>\n",
       "      <td>[MONDO_0009348, EFO_0000640, EFO_0000616, EFO_...</td>\n",
       "      <td>138</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CHEMBL708</td>\n",
       "      <td>[{'disease': 'MONDO_0005090', 'efoName': 'schi...</td>\n",
       "      <td>[MONDO_0005090, EFO_0004269, EFO_0005306, MOND...</td>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CHEMBL1863514</td>\n",
       "      <td>[{'disease': 'EFO_0000220', 'efoName': 'acute ...</td>\n",
       "      <td>[MONDO_0000873, EFO_0000220]</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CHEMBL13209</td>\n",
       "      <td>[{'disease': 'MONDO_0002050', 'efoName': 'depr...</td>\n",
       "      <td>[]</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              id                                        indications  \\\n",
       "0  CHEMBL2146121  [{'disease': 'EFO_0000565', 'efoName': 'leukem...   \n",
       "1  CHEMBL2108738  [{'disease': 'EFO_0005922', 'efoName': 'esopha...   \n",
       "2      CHEMBL708  [{'disease': 'MONDO_0005090', 'efoName': 'schi...   \n",
       "3  CHEMBL1863514  [{'disease': 'EFO_0000220', 'efoName': 'acute ...   \n",
       "4    CHEMBL13209  [{'disease': 'MONDO_0002050', 'efoName': 'depr...   \n",
       "\n",
       "                                 approvedIndications  indicationCount  \n",
       "0                                                 []               42  \n",
       "1  [MONDO_0009348, EFO_0000640, EFO_0000616, EFO_...              138  \n",
       "2  [MONDO_0005090, EFO_0004269, EFO_0005306, MOND...               23  \n",
       "3                       [MONDO_0000873, EFO_0000220]                5  \n",
       "4                                                 []                6  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "Int64Index: 6610 entries, 0 to 36\n",
      "Data columns (total 7 columns):\n",
      " #   Column             Non-Null Count  Dtype \n",
      "---  ------             --------------  ----- \n",
      " 0   actionType         6213 non-null   object\n",
      " 1   mechanismOfAction  6610 non-null   object\n",
      " 2   chemblIds          6610 non-null   object\n",
      " 3   targetName         5849 non-null   object\n",
      " 4   targetType         5849 non-null   object\n",
      " 5   targets            5849 non-null   object\n",
      " 6   references         6610 non-null   object\n",
      "dtypes: object(7)\n",
      "memory usage: 413.1+ KB\n",
      "None\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>actionType</th>\n",
       "      <th>mechanismOfAction</th>\n",
       "      <th>chemblIds</th>\n",
       "      <th>targetName</th>\n",
       "      <th>targetType</th>\n",
       "      <th>targets</th>\n",
       "      <th>references</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>INHIBITOR</td>\n",
       "      <td>Vanilloid receptor inhibitor</td>\n",
       "      <td>[CHEMBL3544953]</td>\n",
       "      <td>Vanilloid receptor</td>\n",
       "      <td>single protein</td>\n",
       "      <td>[ENSG00000196689]</td>\n",
       "      <td>[{'source': 'Other', 'ids': ['http://www.resea...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ANTAGONIST</td>\n",
       "      <td>Serotonin 2a (5-HT2a) receptor antagonist</td>\n",
       "      <td>[CHEMBL1200916, CHEMBL479]</td>\n",
       "      <td>Serotonin 2a (5-HT2a) receptor</td>\n",
       "      <td>single protein</td>\n",
       "      <td>[ENSG00000102468]</td>\n",
       "      <td>[{'source': 'ISBN', 'ids': ['0443-059748 PP. 5...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>INHIBITOR</td>\n",
       "      <td>Polymerase acidic protein inhibitor</td>\n",
       "      <td>[CHEMBL4297215]</td>\n",
       "      <td>Polymerase acidic protein</td>\n",
       "      <td>single protein</td>\n",
       "      <td>[]</td>\n",
       "      <td>[{'source': 'FDA', 'ids': ['https://www.access...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>INHIBITOR</td>\n",
       "      <td>Cyclooxygenase inhibitor</td>\n",
       "      <td>[CHEMBL521]</td>\n",
       "      <td>Cyclooxygenase</td>\n",
       "      <td>protein family</td>\n",
       "      <td>[ENSG00000073756, ENSG00000095303]</td>\n",
       "      <td>[{'source': 'ISBN', 'ids': ['0443-059748 PP. 2...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>None</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>[CHEMBL1334860]</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>[{'source': 'PubMed', 'ids': ['12397207', '149...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   actionType                          mechanismOfAction  \\\n",
       "0   INHIBITOR               Vanilloid receptor inhibitor   \n",
       "1  ANTAGONIST  Serotonin 2a (5-HT2a) receptor antagonist   \n",
       "2   INHIBITOR        Polymerase acidic protein inhibitor   \n",
       "3   INHIBITOR                   Cyclooxygenase inhibitor   \n",
       "4        None                                    Unknown   \n",
       "\n",
       "                    chemblIds                      targetName      targetType  \\\n",
       "0             [CHEMBL3544953]              Vanilloid receptor  single protein   \n",
       "1  [CHEMBL1200916, CHEMBL479]  Serotonin 2a (5-HT2a) receptor  single protein   \n",
       "2             [CHEMBL4297215]       Polymerase acidic protein  single protein   \n",
       "3                 [CHEMBL521]                  Cyclooxygenase  protein family   \n",
       "4             [CHEMBL1334860]                            None            None   \n",
       "\n",
       "                              targets  \\\n",
       "0                   [ENSG00000196689]   \n",
       "1                   [ENSG00000102468]   \n",
       "2                                  []   \n",
       "3  [ENSG00000073756, ENSG00000095303]   \n",
       "4                                None   \n",
       "\n",
       "                                          references  \n",
       "0  [{'source': 'Other', 'ids': ['http://www.resea...  \n",
       "1  [{'source': 'ISBN', 'ids': ['0443-059748 PP. 5...  \n",
       "2  [{'source': 'FDA', 'ids': ['https://www.access...  \n",
       "3  [{'source': 'ISBN', 'ids': ['0443-059748 PP. 2...  \n",
       "4  [{'source': 'PubMed', 'ids': ['12397207', '149...  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "Int64Index: 22274 entries, 0 to 104\n",
      "Data columns (total 16 columns):\n",
      " #   Column               Non-Null Count  Dtype \n",
      "---  ------               --------------  ----- \n",
      " 0   id                   22274 non-null  object\n",
      " 1   code                 22274 non-null  object\n",
      " 2   dbXRefs              22274 non-null  object\n",
      " 3   description          18226 non-null  object\n",
      " 4   name                 22274 non-null  object\n",
      " 5   directLocationIds    12 non-null     object\n",
      " 6   obsoleteTerms        4265 non-null   object\n",
      " 7   parents              22274 non-null  object\n",
      " 8   sko                  1 non-null      object\n",
      " 9   synonyms             14593 non-null  object\n",
      " 10  ancestors            22274 non-null  object\n",
      " 11  descendants          22274 non-null  object\n",
      " 12  children             22274 non-null  object\n",
      " 13  therapeuticAreas     22274 non-null  object\n",
      " 14  indirectLocationIds  34 non-null     object\n",
      " 15  ontology             22274 non-null  object\n",
      "dtypes: object(16)\n",
      "memory usage: 2.9+ MB\n",
      "None\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>code</th>\n",
       "      <th>dbXRefs</th>\n",
       "      <th>description</th>\n",
       "      <th>name</th>\n",
       "      <th>directLocationIds</th>\n",
       "      <th>obsoleteTerms</th>\n",
       "      <th>parents</th>\n",
       "      <th>sko</th>\n",
       "      <th>synonyms</th>\n",
       "      <th>ancestors</th>\n",
       "      <th>descendants</th>\n",
       "      <th>children</th>\n",
       "      <th>therapeuticAreas</th>\n",
       "      <th>indirectLocationIds</th>\n",
       "      <th>ontology</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>EFO_0000224</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0000224</td>\n",
       "      <td>[NCIT:C3182, MONDO:0012883, UMLS:C0023487, EFO...</td>\n",
       "      <td>Acute promyelocytic leukemia (APL) is an aggre...</td>\n",
       "      <td>acute promyelocytic leukemia</td>\n",
       "      <td>None</td>\n",
       "      <td>[EFO_0000202]</td>\n",
       "      <td>[EFO_0000222]</td>\n",
       "      <td>None</td>\n",
       "      <td>{'hasBroadSynonym': None, 'hasExactSynonym': [...</td>\n",
       "      <td>[MONDO_0004643, EFO_0004260, OTAR_0000006, EFO...</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[OTAR_0000006, EFO_0005803, MONDO_0045024, EFO...</td>\n",
       "      <td>None</td>\n",
       "      <td>{'isTherapeuticArea': False, 'leaf': True, 'so...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>EFO_0000384</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0000384</td>\n",
       "      <td>[MedDRA:10011401, MONDO:0005011, NCIt:C27837, ...</td>\n",
       "      <td>A chronic transmural inflammation that may inv...</td>\n",
       "      <td>Crohn's disease</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>[EFO_0003767]</td>\n",
       "      <td>None</td>\n",
       "      <td>{'hasBroadSynonym': None, 'hasExactSynonym': [...</td>\n",
       "      <td>[EFO_0005140, EFO_0009431, OTAR_0000018, EFO_0...</td>\n",
       "      <td>[MONDO_0000709, EFO_0005622, EFO_0005624, EFO_...</td>\n",
       "      <td>[EFO_0005622, EFO_0005625, EFO_0005627, EFO_00...</td>\n",
       "      <td>[OTAR_0000018, EFO_0000540, EFO_0010282]</td>\n",
       "      <td>None</td>\n",
       "      <td>{'isTherapeuticArea': False, 'leaf': False, 's...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>EFO_0001421</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0001421</td>\n",
       "      <td>[DOID:409, UMLS:C0023895, NCIt:C3196, ICD9:573...</td>\n",
       "      <td>Any disease or dysfunction of the liver and th...</td>\n",
       "      <td>liver disease</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>[EFO_0001379, EFO_0010284]</td>\n",
       "      <td>None</td>\n",
       "      <td>{'hasBroadSynonym': None, 'hasExactSynonym': [...</td>\n",
       "      <td>[EFO_0001379, EFO_0010284, EFO_0010282]</td>\n",
       "      <td>[MONDO_0008967, MONDO_0008966, MONDO_0000447, ...</td>\n",
       "      <td>[MONDO_0000447, MONDO_0002405, MONDO_0002520, ...</td>\n",
       "      <td>[EFO_0001379, EFO_0010282]</td>\n",
       "      <td>None</td>\n",
       "      <td>{'isTherapeuticArea': False, 'leaf': False, 's...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>EFO_0004328</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0004328</td>\n",
       "      <td>[MeSH:D005080, MedDRA:10015652]</td>\n",
       "      <td>Controlled physical activity, more strenuous t...</td>\n",
       "      <td>exercise test</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>[EFO_0004311]</td>\n",
       "      <td>None</td>\n",
       "      <td>{'hasBroadSynonym': None, 'hasExactSynonym': [...</td>\n",
       "      <td>[EFO_0001444, EFO_0005278, EFO_0004311, EFO_00...</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[EFO_0001444]</td>\n",
       "      <td>None</td>\n",
       "      <td>{'isTherapeuticArea': False, 'leaf': True, 'so...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>EFO_0004554</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0004554</td>\n",
       "      <td>[]</td>\n",
       "      <td>Is a quantification of some aspect of the geno...</td>\n",
       "      <td>genomic measurement</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>[EFO_0001444]</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>[EFO_0001444]</td>\n",
       "      <td>[EFO_0009861, EFO_0009860, EFO_0007783, EFO_00...</td>\n",
       "      <td>[EFO_0000513, EFO_0004505, EFO_0004798, EFO_00...</td>\n",
       "      <td>[EFO_0001444]</td>\n",
       "      <td>None</td>\n",
       "      <td>{'isTherapeuticArea': False, 'leaf': False, 's...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            id                                  code  \\\n",
       "0  EFO_0000224  http://www.ebi.ac.uk/efo/EFO_0000224   \n",
       "1  EFO_0000384  http://www.ebi.ac.uk/efo/EFO_0000384   \n",
       "2  EFO_0001421  http://www.ebi.ac.uk/efo/EFO_0001421   \n",
       "3  EFO_0004328  http://www.ebi.ac.uk/efo/EFO_0004328   \n",
       "4  EFO_0004554  http://www.ebi.ac.uk/efo/EFO_0004554   \n",
       "\n",
       "                                             dbXRefs  \\\n",
       "0  [NCIT:C3182, MONDO:0012883, UMLS:C0023487, EFO...   \n",
       "1  [MedDRA:10011401, MONDO:0005011, NCIt:C27837, ...   \n",
       "2  [DOID:409, UMLS:C0023895, NCIt:C3196, ICD9:573...   \n",
       "3                    [MeSH:D005080, MedDRA:10015652]   \n",
       "4                                                 []   \n",
       "\n",
       "                                         description  \\\n",
       "0  Acute promyelocytic leukemia (APL) is an aggre...   \n",
       "1  A chronic transmural inflammation that may inv...   \n",
       "2  Any disease or dysfunction of the liver and th...   \n",
       "3  Controlled physical activity, more strenuous t...   \n",
       "4  Is a quantification of some aspect of the geno...   \n",
       "\n",
       "                           name directLocationIds  obsoleteTerms  \\\n",
       "0  acute promyelocytic leukemia              None  [EFO_0000202]   \n",
       "1               Crohn's disease              None           None   \n",
       "2                 liver disease              None           None   \n",
       "3                 exercise test              None           None   \n",
       "4           genomic measurement              None           None   \n",
       "\n",
       "                      parents   sko  \\\n",
       "0               [EFO_0000222]  None   \n",
       "1               [EFO_0003767]  None   \n",
       "2  [EFO_0001379, EFO_0010284]  None   \n",
       "3               [EFO_0004311]  None   \n",
       "4               [EFO_0001444]  None   \n",
       "\n",
       "                                            synonyms  \\\n",
       "0  {'hasBroadSynonym': None, 'hasExactSynonym': [...   \n",
       "1  {'hasBroadSynonym': None, 'hasExactSynonym': [...   \n",
       "2  {'hasBroadSynonym': None, 'hasExactSynonym': [...   \n",
       "3  {'hasBroadSynonym': None, 'hasExactSynonym': [...   \n",
       "4                                               None   \n",
       "\n",
       "                                           ancestors  \\\n",
       "0  [MONDO_0004643, EFO_0004260, OTAR_0000006, EFO...   \n",
       "1  [EFO_0005140, EFO_0009431, OTAR_0000018, EFO_0...   \n",
       "2            [EFO_0001379, EFO_0010284, EFO_0010282]   \n",
       "3  [EFO_0001444, EFO_0005278, EFO_0004311, EFO_00...   \n",
       "4                                      [EFO_0001444]   \n",
       "\n",
       "                                         descendants  \\\n",
       "0                                                 []   \n",
       "1  [MONDO_0000709, EFO_0005622, EFO_0005624, EFO_...   \n",
       "2  [MONDO_0008967, MONDO_0008966, MONDO_0000447, ...   \n",
       "3                                                 []   \n",
       "4  [EFO_0009861, EFO_0009860, EFO_0007783, EFO_00...   \n",
       "\n",
       "                                            children  \\\n",
       "0                                                 []   \n",
       "1  [EFO_0005622, EFO_0005625, EFO_0005627, EFO_00...   \n",
       "2  [MONDO_0000447, MONDO_0002405, MONDO_0002520, ...   \n",
       "3                                                 []   \n",
       "4  [EFO_0000513, EFO_0004505, EFO_0004798, EFO_00...   \n",
       "\n",
       "                                    therapeuticAreas indirectLocationIds  \\\n",
       "0  [OTAR_0000006, EFO_0005803, MONDO_0045024, EFO...                None   \n",
       "1           [OTAR_0000018, EFO_0000540, EFO_0010282]                None   \n",
       "2                         [EFO_0001379, EFO_0010282]                None   \n",
       "3                                      [EFO_0001444]                None   \n",
       "4                                      [EFO_0001444]                None   \n",
       "\n",
       "                                            ontology  \n",
       "0  {'isTherapeuticArea': False, 'leaf': True, 'so...  \n",
       "1  {'isTherapeuticArea': False, 'leaf': False, 's...  \n",
       "2  {'isTherapeuticArea': False, 'leaf': False, 's...  \n",
       "3  {'isTherapeuticArea': False, 'leaf': True, 'so...  \n",
       "4  {'isTherapeuticArea': False, 'leaf': False, 's...  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "Int64Index: 62678 entries, 0 to 332\n",
      "Data columns (total 28 columns):\n",
      " #   Column                Non-Null Count  Dtype \n",
      "---  ------                --------------  ----- \n",
      " 0   id                    62678 non-null  object\n",
      " 1   approvedSymbol        62678 non-null  object\n",
      " 2   biotype               62678 non-null  object\n",
      " 3   transcriptIds         62678 non-null  object\n",
      " 4   canonicalTranscript   61418 non-null  object\n",
      " 5   canonicalExons        61381 non-null  object\n",
      " 6   genomicLocation       62678 non-null  object\n",
      " 7   alternativeGenes      1086 non-null   object\n",
      " 8   approvedName          62678 non-null  object\n",
      " 9   go                    20904 non-null  object\n",
      " 10  hallmarks             351 non-null    object\n",
      " 11  synonyms              62678 non-null  object\n",
      " 12  symbolSynonyms        62678 non-null  object\n",
      " 13  nameSynonyms          62678 non-null  object\n",
      " 14  functionDescriptions  20135 non-null  object\n",
      " 15  subcellularLocations  20135 non-null  object\n",
      " 16  targetClass           4648 non-null   object\n",
      " 17  obsoleteSymbols       62678 non-null  object\n",
      " 18  obsoleteNames         62678 non-null  object\n",
      " 19  constraint            19099 non-null  object\n",
      " 20  tep                   41 non-null     object\n",
      " 21  proteinIds            21064 non-null  object\n",
      " 22  dbXrefs               62678 non-null  object\n",
      " 23  chemicalProbes        572 non-null    object\n",
      " 24  homologues            25417 non-null  object\n",
      " 25  tractability          19471 non-null  object\n",
      " 26  safetyLiabilities     456 non-null    object\n",
      " 27  pathways              11010 non-null  object\n",
      "dtypes: object(28)\n",
      "memory usage: 13.9+ MB\n",
      "None\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>approvedSymbol</th>\n",
       "      <th>biotype</th>\n",
       "      <th>transcriptIds</th>\n",
       "      <th>canonicalTranscript</th>\n",
       "      <th>canonicalExons</th>\n",
       "      <th>genomicLocation</th>\n",
       "      <th>alternativeGenes</th>\n",
       "      <th>approvedName</th>\n",
       "      <th>go</th>\n",
       "      <th>...</th>\n",
       "      <th>obsoleteNames</th>\n",
       "      <th>constraint</th>\n",
       "      <th>tep</th>\n",
       "      <th>proteinIds</th>\n",
       "      <th>dbXrefs</th>\n",
       "      <th>chemicalProbes</th>\n",
       "      <th>homologues</th>\n",
       "      <th>tractability</th>\n",
       "      <th>safetyLiabilities</th>\n",
       "      <th>pathways</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000011451</td>\n",
       "      <td>WIZ</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>[ENST00000389282, ENST00000596159, ENST0000026...</td>\n",
       "      <td>{'id': 'ENST00000673675', 'chromosome': '19', ...</td>\n",
       "      <td>[15436806, 15437129, 15426982, 15427533, 15425...</td>\n",
       "      <td>{'chromosome': '19', 'start': 15419978, 'end':...</td>\n",
       "      <td>None</td>\n",
       "      <td>WIZ zinc finger</td>\n",
       "      <td>[{'id': 'GO:0006357', 'source': 'PMID:21873635...</td>\n",
       "      <td>...</td>\n",
       "      <td>[{'label': 'widely interspaced zinc finger mot...</td>\n",
       "      <td>[{'constraintType': 'syn', 'score': 0.31986999...</td>\n",
       "      <td>None</td>\n",
       "      <td>[{'id': 'O95785', 'source': 'uniprot_swissprot...</td>\n",
       "      <td>[{'id': '30917', 'source': 'HGNC'}, {'id': 'IP...</td>\n",
       "      <td>None</td>\n",
       "      <td>[{'speciesId': '9606', 'speciesName': 'Human',...</td>\n",
       "      <td>[{'modality': 'SM', 'id': 'Approved Drug', 'va...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000023902</td>\n",
       "      <td>PLEKHO1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>[ENST00000485470, ENST00000492304, ENST0000060...</td>\n",
       "      <td>{'id': 'ENST00000369124', 'chromosome': '1', '...</td>\n",
       "      <td>[150158819, 150160065, 150149916, 150150287, 1...</td>\n",
       "      <td>{'chromosome': '1', 'start': 150149183, 'end':...</td>\n",
       "      <td>None</td>\n",
       "      <td>pleckstrin homology domain containing O1</td>\n",
       "      <td>[{'id': 'GO:0007520', 'source': 'GO_REF:000010...</td>\n",
       "      <td>...</td>\n",
       "      <td>[{'label': 'pleckstrin homology domain contain...</td>\n",
       "      <td>[{'constraintType': 'syn', 'score': -0.2883000...</td>\n",
       "      <td>None</td>\n",
       "      <td>[{'id': 'Q53GL0', 'source': 'uniprot_swissprot...</td>\n",
       "      <td>[{'id': '24310', 'source': 'HGNC'}, {'id': '3A...</td>\n",
       "      <td>None</td>\n",
       "      <td>[{'speciesId': '9606', 'speciesName': 'Human',...</td>\n",
       "      <td>[{'modality': 'SM', 'id': 'Approved Drug', 'va...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000035862</td>\n",
       "      <td>TIMP2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>[ENST00000536189, ENST00000586057, ENST0000058...</td>\n",
       "      <td>{'id': 'ENST00000262768', 'chromosome': '17', ...</td>\n",
       "      <td>[78857522, 78857646, 78924959, 78925387, 78852...</td>\n",
       "      <td>{'chromosome': '17', 'start': 78852977, 'end':...</td>\n",
       "      <td>None</td>\n",
       "      <td>TIMP metallopeptidase inhibitor 2</td>\n",
       "      <td>[{'id': 'GO:0005576', 'source': 'Reactome:R-HS...</td>\n",
       "      <td>...</td>\n",
       "      <td>[{'label': 'tissue inhibitor of metalloprotein...</td>\n",
       "      <td>[{'constraintType': 'syn', 'score': -0.0323360...</td>\n",
       "      <td>None</td>\n",
       "      <td>[{'id': 'P16035', 'source': 'uniprot_swissprot...</td>\n",
       "      <td>[{'id': '11821', 'source': 'HGNC'}, {'id': '1B...</td>\n",
       "      <td>None</td>\n",
       "      <td>[{'speciesId': '9606', 'speciesName': 'Human',...</td>\n",
       "      <td>[{'modality': 'SM', 'id': 'Approved Drug', 'va...</td>\n",
       "      <td>[{'event': 'regulation of gene expression', 'e...</td>\n",
       "      <td>[{'pathwayId': 'R-HSA-1592389', 'pathway': 'Ac...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000042813</td>\n",
       "      <td>ZPBP</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>[ENST00000046087, ENST00000465922, ENST0000041...</td>\n",
       "      <td>{'id': 'ENST00000046087', 'chromosome': '7', '...</td>\n",
       "      <td>[50018240, 50018316, 50031092, 50031310, 50057...</td>\n",
       "      <td>{'chromosome': '7', 'start': 49850421, 'end': ...</td>\n",
       "      <td>None</td>\n",
       "      <td>zona pellucida binding protein</td>\n",
       "      <td>[{'id': 'GO:0001669', 'source': 'PMID:21873635...</td>\n",
       "      <td>...</td>\n",
       "      <td>[]</td>\n",
       "      <td>[{'constraintType': 'syn', 'score': -1.2680000...</td>\n",
       "      <td>None</td>\n",
       "      <td>[{'id': 'Q9BS86', 'source': 'uniprot_swissprot...</td>\n",
       "      <td>[{'id': '15662', 'source': 'HGNC'}, {'id': 'IP...</td>\n",
       "      <td>None</td>\n",
       "      <td>[{'speciesId': '9606', 'speciesName': 'Human',...</td>\n",
       "      <td>[{'modality': 'SM', 'id': 'Approved Drug', 'va...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000065457</td>\n",
       "      <td>ADAT1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>[ENST00000568001, ENST00000568510, ENST0000030...</td>\n",
       "      <td>{'id': 'ENST00000564657', 'chromosome': '16', ...</td>\n",
       "      <td>[75622403, 75623281, 75620631, 75620820, 75620...</td>\n",
       "      <td>{'chromosome': '16', 'start': 75596868, 'end':...</td>\n",
       "      <td>None</td>\n",
       "      <td>adenosine deaminase tRNA specific 1</td>\n",
       "      <td>[{'id': 'GO:0003723', 'source': 'PMID:10430867...</td>\n",
       "      <td>...</td>\n",
       "      <td>[]</td>\n",
       "      <td>[{'constraintType': 'syn', 'score': -1.4117000...</td>\n",
       "      <td>None</td>\n",
       "      <td>[{'id': 'Q9BUB4', 'source': 'uniprot_swissprot...</td>\n",
       "      <td>[{'id': '228', 'source': 'HGNC'}, {'id': 'R-HS...</td>\n",
       "      <td>None</td>\n",
       "      <td>[{'speciesId': '9606', 'speciesName': 'Human',...</td>\n",
       "      <td>[{'modality': 'SM', 'id': 'Approved Drug', 'va...</td>\n",
       "      <td>None</td>\n",
       "      <td>[{'pathwayId': 'R-HSA-6782315', 'pathway': 'tR...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 28 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                id approvedSymbol         biotype  \\\n",
       "0  ENSG00000011451            WIZ  protein_coding   \n",
       "1  ENSG00000023902        PLEKHO1  protein_coding   \n",
       "2  ENSG00000035862          TIMP2  protein_coding   \n",
       "3  ENSG00000042813           ZPBP  protein_coding   \n",
       "4  ENSG00000065457          ADAT1  protein_coding   \n",
       "\n",
       "                                       transcriptIds  \\\n",
       "0  [ENST00000389282, ENST00000596159, ENST0000026...   \n",
       "1  [ENST00000485470, ENST00000492304, ENST0000060...   \n",
       "2  [ENST00000536189, ENST00000586057, ENST0000058...   \n",
       "3  [ENST00000046087, ENST00000465922, ENST0000041...   \n",
       "4  [ENST00000568001, ENST00000568510, ENST0000030...   \n",
       "\n",
       "                                 canonicalTranscript  \\\n",
       "0  {'id': 'ENST00000673675', 'chromosome': '19', ...   \n",
       "1  {'id': 'ENST00000369124', 'chromosome': '1', '...   \n",
       "2  {'id': 'ENST00000262768', 'chromosome': '17', ...   \n",
       "3  {'id': 'ENST00000046087', 'chromosome': '7', '...   \n",
       "4  {'id': 'ENST00000564657', 'chromosome': '16', ...   \n",
       "\n",
       "                                      canonicalExons  \\\n",
       "0  [15436806, 15437129, 15426982, 15427533, 15425...   \n",
       "1  [150158819, 150160065, 150149916, 150150287, 1...   \n",
       "2  [78857522, 78857646, 78924959, 78925387, 78852...   \n",
       "3  [50018240, 50018316, 50031092, 50031310, 50057...   \n",
       "4  [75622403, 75623281, 75620631, 75620820, 75620...   \n",
       "\n",
       "                                     genomicLocation alternativeGenes  \\\n",
       "0  {'chromosome': '19', 'start': 15419978, 'end':...             None   \n",
       "1  {'chromosome': '1', 'start': 150149183, 'end':...             None   \n",
       "2  {'chromosome': '17', 'start': 78852977, 'end':...             None   \n",
       "3  {'chromosome': '7', 'start': 49850421, 'end': ...             None   \n",
       "4  {'chromosome': '16', 'start': 75596868, 'end':...             None   \n",
       "\n",
       "                               approvedName  \\\n",
       "0                           WIZ zinc finger   \n",
       "1  pleckstrin homology domain containing O1   \n",
       "2         TIMP metallopeptidase inhibitor 2   \n",
       "3            zona pellucida binding protein   \n",
       "4       adenosine deaminase tRNA specific 1   \n",
       "\n",
       "                                                  go  ...  \\\n",
       "0  [{'id': 'GO:0006357', 'source': 'PMID:21873635...  ...   \n",
       "1  [{'id': 'GO:0007520', 'source': 'GO_REF:000010...  ...   \n",
       "2  [{'id': 'GO:0005576', 'source': 'Reactome:R-HS...  ...   \n",
       "3  [{'id': 'GO:0001669', 'source': 'PMID:21873635...  ...   \n",
       "4  [{'id': 'GO:0003723', 'source': 'PMID:10430867...  ...   \n",
       "\n",
       "                                       obsoleteNames  \\\n",
       "0  [{'label': 'widely interspaced zinc finger mot...   \n",
       "1  [{'label': 'pleckstrin homology domain contain...   \n",
       "2  [{'label': 'tissue inhibitor of metalloprotein...   \n",
       "3                                                 []   \n",
       "4                                                 []   \n",
       "\n",
       "                                          constraint   tep  \\\n",
       "0  [{'constraintType': 'syn', 'score': 0.31986999...  None   \n",
       "1  [{'constraintType': 'syn', 'score': -0.2883000...  None   \n",
       "2  [{'constraintType': 'syn', 'score': -0.0323360...  None   \n",
       "3  [{'constraintType': 'syn', 'score': -1.2680000...  None   \n",
       "4  [{'constraintType': 'syn', 'score': -1.4117000...  None   \n",
       "\n",
       "                                          proteinIds  \\\n",
       "0  [{'id': 'O95785', 'source': 'uniprot_swissprot...   \n",
       "1  [{'id': 'Q53GL0', 'source': 'uniprot_swissprot...   \n",
       "2  [{'id': 'P16035', 'source': 'uniprot_swissprot...   \n",
       "3  [{'id': 'Q9BS86', 'source': 'uniprot_swissprot...   \n",
       "4  [{'id': 'Q9BUB4', 'source': 'uniprot_swissprot...   \n",
       "\n",
       "                                             dbXrefs chemicalProbes  \\\n",
       "0  [{'id': '30917', 'source': 'HGNC'}, {'id': 'IP...           None   \n",
       "1  [{'id': '24310', 'source': 'HGNC'}, {'id': '3A...           None   \n",
       "2  [{'id': '11821', 'source': 'HGNC'}, {'id': '1B...           None   \n",
       "3  [{'id': '15662', 'source': 'HGNC'}, {'id': 'IP...           None   \n",
       "4  [{'id': '228', 'source': 'HGNC'}, {'id': 'R-HS...           None   \n",
       "\n",
       "                                          homologues  \\\n",
       "0  [{'speciesId': '9606', 'speciesName': 'Human',...   \n",
       "1  [{'speciesId': '9606', 'speciesName': 'Human',...   \n",
       "2  [{'speciesId': '9606', 'speciesName': 'Human',...   \n",
       "3  [{'speciesId': '9606', 'speciesName': 'Human',...   \n",
       "4  [{'speciesId': '9606', 'speciesName': 'Human',...   \n",
       "\n",
       "                                        tractability  \\\n",
       "0  [{'modality': 'SM', 'id': 'Approved Drug', 'va...   \n",
       "1  [{'modality': 'SM', 'id': 'Approved Drug', 'va...   \n",
       "2  [{'modality': 'SM', 'id': 'Approved Drug', 'va...   \n",
       "3  [{'modality': 'SM', 'id': 'Approved Drug', 'va...   \n",
       "4  [{'modality': 'SM', 'id': 'Approved Drug', 'va...   \n",
       "\n",
       "                                   safetyLiabilities  \\\n",
       "0                                               None   \n",
       "1                                               None   \n",
       "2  [{'event': 'regulation of gene expression', 'e...   \n",
       "3                                               None   \n",
       "4                                               None   \n",
       "\n",
       "                                            pathways  \n",
       "0                                               None  \n",
       "1                                               None  \n",
       "2  [{'pathwayId': 'R-HSA-1592389', 'pathway': 'Ac...  \n",
       "3                                               None  \n",
       "4  [{'pathwayId': 'R-HSA-6782315', 'pathway': 'tR...  \n",
       "\n",
       "[5 rows x 28 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "for df in [drug_df, indication_df, moa_df, disease_df, target_df]:\n",
    "    print(df.info())\n",
    "    display(df.head(5))"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1.2 Subset data\n",
    "\n",
    "I want to limit the exercise to a small amount of data to test that the model is functional without having to embed the whole OT universe.\n",
    "To do so:\n",
    "- I will select a random sample of 100 rows per dataset.\n",
    "- I will use records that are as complete as possible (no nulls in any of the columns)\n",
    "  - This will not always be possible (drug_df, disease_df, target_df) have at least one null value."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [],
   "source": [
    "def subset_shuffle_data(df: pd.DataFrame) -> pd.DataFrame:\n",
    "    if len(df.dropna(axis=0, how=\"any\")) >= 100:\n",
    "        return df.dropna(axis=0, how=\"any\").sample(100).reset_index(drop=True)\n",
    "    return df.sample(100).reset_index(drop=True)\n",
    "\n",
    "drug_subset_df = subset_shuffle_data(drug_df)\n",
    "indication_subset_df = subset_shuffle_data(indication_df)\n",
    "moa_subset_df = subset_shuffle_data(moa_df)\n",
    "disease_subset_df = subset_shuffle_data(disease_df)\n",
    "target_subset_df = subset_shuffle_data(target_df)\n",
    "\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>actionType</th>\n",
       "      <th>mechanismOfAction</th>\n",
       "      <th>chemblIds</th>\n",
       "      <th>targetName</th>\n",
       "      <th>targetType</th>\n",
       "      <th>targets</th>\n",
       "      <th>references</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>INHIBITOR</td>\n",
       "      <td>Tubulin inhibitor</td>\n",
       "      <td>[CHEMBL3833306]</td>\n",
       "      <td>Tubulin</td>\n",
       "      <td>protein complex group</td>\n",
       "      <td>[ENSG00000188229, ENSG00000101162, ENSG0000019...</td>\n",
       "      <td>[{'source': 'PubMed', 'ids': ['27518442', '283...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ANTAGONIST</td>\n",
       "      <td>Dopamine D2 receptor antagonist</td>\n",
       "      <td>[CHEMBL28218]</td>\n",
       "      <td>Dopamine D2 receptor</td>\n",
       "      <td>single protein</td>\n",
       "      <td>[ENSG00000149295]</td>\n",
       "      <td>[{'source': 'PubMed', 'ids': ['17362435', '215...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AGONIST</td>\n",
       "      <td>Glucocorticoid receptor agonist</td>\n",
       "      <td>[CHEMBL1549]</td>\n",
       "      <td>Glucocorticoid receptor</td>\n",
       "      <td>single protein</td>\n",
       "      <td>[ENSG00000113580]</td>\n",
       "      <td>[{'source': 'Wikipedia', 'ids': ['Glucocortico...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   actionType                mechanismOfAction        chemblIds  \\\n",
       "0   INHIBITOR                Tubulin inhibitor  [CHEMBL3833306]   \n",
       "1  ANTAGONIST  Dopamine D2 receptor antagonist    [CHEMBL28218]   \n",
       "2     AGONIST  Glucocorticoid receptor agonist     [CHEMBL1549]   \n",
       "\n",
       "                targetName             targetType  \\\n",
       "0                  Tubulin  protein complex group   \n",
       "1     Dopamine D2 receptor         single protein   \n",
       "2  Glucocorticoid receptor         single protein   \n",
       "\n",
       "                                             targets  \\\n",
       "0  [ENSG00000188229, ENSG00000101162, ENSG0000019...   \n",
       "1                                  [ENSG00000149295]   \n",
       "2                                  [ENSG00000113580]   \n",
       "\n",
       "                                          references  \n",
       "0  [{'source': 'PubMed', 'ids': ['27518442', '283...  \n",
       "1  [{'source': 'PubMed', 'ids': ['17362435', '215...  \n",
       "2  [{'source': 'Wikipedia', 'ids': ['Glucocortico...  "
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "moa_subset_df.head(3)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2. Generate embeddings\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[ 0.02768615, -0.04751104, -0.01583998, ..., -0.0343785 ,\n",
       "         0.0130953 ,  0.0593827 ],\n",
       "       [ 0.00734502, -0.0574134 , -0.00358513, ..., -0.04301181,\n",
       "         0.01122494,  0.08128042]], dtype=float32)"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.encode([\"This is a test\", \"This is another test\"])\n",
    "\n",
    "# nice, i get a vector of size 768 (same parameters as the T5 model)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "I have 2 ideas to query the model:\n",
    "\n",
    "1. Provide all the data in the prompt and the question I want to ask. For this I don't know if I have to be explicit every time (e.g. in which columns I have drug IDs?) or if I can just ask for all the relationships it can find. This is kind of a zero shot learning approach.\n",
    "\n",
    "2. Index each column and then ask the model to find the relationships between them by using cosine similarities. This has the drawback of having to be explicit.\n",
    "\n",
    "Let's go for the harder one first (#1)\n",
    "### 2.1 Provide all data in the prompt by converting the tables to text"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'\\nBased on the information provided, it seems that the common ID between the dataframes is the CHEMBL ID. Here are the columns in each dataframe that contain drug IDs:\\n\\nDrug_df Dataframe:\\n\\nid\\nparentId\\nchildChemblIds\\ncrossReferences\\nIndications_df dataframe:\\n\\nid\\nindications <--- this is wrong\\nMechanism_of_action_df Dataframe:\\n\\nchemblIds\\n\\n'"
      ]
     },
     "execution_count": 84,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# WIP: I have been testing this solution with Cohere's API without great results, I am still trying to figure out what is the best prompt.\n",
    "\n",
    "\"\"\"\n",
    "Sample output (prompt was one record of each datasets, question was \"Give me all columns in each dataframe that contain drug IDs to draw relationships between the data:\"):\n",
    "To draw relationships between the data, you can use the join() function. For example, you can join the Drug_df and Indications_df dataframes on the drug ID column to get a dataframe with the combined columns.\n",
    "\n",
    "# Create a new dataframe with the joined columns\n",
    "drug_indications = pd.join(Drug_df, Indications_df, on='drug_id')\n",
    "\n",
    "This will give you a dataframe with the combined columns, including the drug ID column from both dataframes. You can then use this dataframe to draw relationships between the data.\n",
    "\"\"\"\n",
    "\n",
    "# However, for the same promt ChatGPT gave me almost exactly what I wanted (I even hadn't realised there are drug IDs in crossReferences too!):\n",
    "\"\"\"\n",
    "Based on the information provided, it seems that the common ID between the dataframes is the CHEMBL ID. Here are the columns in each dataframe that contain drug IDs:\n",
    "\n",
    "Drug_df Dataframe:\n",
    "\n",
    "id\n",
    "parentId\n",
    "childChemblIds\n",
    "crossReferences\n",
    "Indications_df dataframe:\n",
    "\n",
    "id\n",
    "indications <--- this is wrong\n",
    "Mechanism_of_action_df Dataframe:\n",
    "\n",
    "chemblIds\n",
    "\n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2.2 Vectorise each column of each dataset and compute similarities\n",
    "\n",
    "My dictionary will have three keys:\n",
    "\n",
    "- id: following the pattern \"{dataframe_name}_{column_name}\"\n",
    "- content: each value of the column in a single string format\n",
    "- embedding: the embedding of the content"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "dict_keys(['test_id', 'test_indications', 'test_approvedIndications', 'test_indicationCount'])\n",
      "['CHEMBL2109540', 'CHEMBL4297812', 'CHEMBL2109322', 'CHEMBL4297809', 'CHEMBL4594404', 'CHEMBL2105665', 'CHEMBL2108327', 'CHEMBL577711', 'CHEMBL4297444', 'CHEMBL1643', 'CHEMBL1493', 'CHEMBL3707289', 'CHEMBL4299940', 'CHEMBL2392545', 'CHEMBL4594547', 'CHEMBL1214192', 'CHEMBL1241348', 'CHEMBL218427', 'CHEMBL65794', 'CHEMBL2151437', 'CHEMBL1087', 'CHEMBL2108791', 'CHEMBL211538', 'CHEMBL2109249', 'CHEMBL1457', 'CHEMBL1082508', 'CHEMBL516', 'CHEMBL3039550', 'CHEMBL1741078', 'CHEMBL494397', 'CHEMBL1201630', 'CHEMBL1743260', 'CHEMBL1201200', 'CHEMBL294199', 'CHEMBL2110603', 'CHEMBL2107884', 'CHEMBL2109546', 'CHEMBL2165224', 'CHEMBL4297283', 'CHEMBL2111110', 'CHEMBL2108308', 'CHEMBL37676', 'CHEMBL294951', 'CHEMBL160', 'CHEMBL12610', 'CHEMBL4650349', 'CHEMBL113313', 'CHEMBL334966', 'CHEMBL1472989', 'CHEMBL647', 'CHEMBL2220486', 'CHEMBL1086440', 'CHEMBL3545013', 'CHEMBL2109641', 'CHEMBL3545298', 'CHEMBL891', 'CHEMBL4297107', 'CHEMBL1556', 'CHEMBL207413', 'CHEMBL2103764', 'CHEMBL4594348', 'CHEMBL57', 'CHEMBL1213327', 'CHEMBL161', 'CHEMBL2109492', 'CHEMBL473417', 'CHEMBL2109583', 'CHEMBL4650272', 'CHEMBL1630578', 'CHEMBL4068611', 'CHEMBL4594474', 'CHEMBL4297790', 'CHEMBL3137341', 'CHEMBL2109465', 'CHEMBL1201570', 'CHEMBL1328898', 'CHEMBL35', 'CHEMBL2103875', 'CHEMBL207456', 'CHEMBL2107962', 'CHEMBL19236', 'CHEMBL2111097', 'CHEMBL2110849', 'CHEMBL188952', 'CHEMBL1201759', 'CHEMBL4298001', 'CHEMBL4650436', 'CHEMBL4594495', 'CHEMBL4297641', 'CHEMBL513301', 'CHEMBL452', 'CHEMBL1397', 'CHEMBL4297480', 'CHEMBL1200848', 'CHEMBL1200370', 'CHEMBL2111290', 'CHEMBL1257070', 'CHEMBL4651180', 'CHEMBL57242', 'CHEMBL501680']\n"
     ]
    }
   ],
   "source": [
    "test_df = indication_subset_df.copy()\n",
    "test_df.columns = [f\"test_{col}\" for col in test_df.columns]\n",
    "\n",
    "result_dict = {col: test_df[col].astype(str).tolist() for col in test_df.columns}\n",
    "for k, v in result_dict.items():\n",
    "    result_dict[k] = str(v)\n",
    "\n",
    "print(result_dict.keys())\n",
    "print(result_dict[\"test_id\"])\n",
    "# This is what I want, but i need to review the array columns, converting them into strings introduces weird characters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dict_keys(['test_id', 'test_indications', 'test_approvedIndications', 'test_indicationCount'])"
      ]
     },
     "execution_count": 88,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def populate_embeddings_dict(df: pd.DataFrame, embeddings_dict: dict) -> dict:\n",
    "    # add prefix to column name\n",
    "    df.columns = [f\"test_{col}\" for col in test_df.columns]\n",
    "    result_dict = {col: test_df[col].astype(str).tolist() for col in test_df.columns}\n",
    "    for k, v in result_dict.items():\n",
    "        result_dict[k] = str(v)\n",
    "    embeddings_dict |= result_dict\n",
    "    return embeddings_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.15"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "355034e2c8f3d095a6c50a22c9a577e5665f0d4bc6ffc9f89afa96a0b829f20a"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
